src/sync.rs

```diff
@@ -74,58 +75,137 @@ impl SyncSession {
+                address: src_addr.ip().to_string(),
+                data: buf_out,
+            })
+        } else {
+            Err(SnmpError::ReceiveError)
+        }
+    }
+
     pub fn get<T>(&mut self, names: &[T]) -> SnmpResult<SnmpPdu>
     where
         T: AsRef<[u32]>,
     {
         let req_id = self.req_id.0;
-        pdu::build_get(self.community.as_slice(), req_id, names, &mut self.send_pdu)?;
-        let recv_len = Self::send_and_recv(&self.socket, &self.send_pdu, &mut self.recv_buf[..])?;
         self.req_id += Wrapping(1);
-        let pdu_bytes = &self.recv_buf[..recv_len];
-        handle_response(req_id, self.community.as_slice(), pdu_bytes.into())
+
+        pdu::build_get(self.community.as_slice(), req_id, names, &mut self.send_pdu)?;
+
+        let response = self.send_and_recv(&self.send_pdu)?;
+
+        handle_response(req_id, self.community.as_slice(), response.data.as_slice())
     }
 
     pub fn getnext(&mut self, name: &[u32]) -> SnmpResult<SnmpPdu> {
         let req_id = self.req_id.0;
+        self.req_id += Wrapping(1);
+
         pdu::build_getnext(self.community.as_slice(), req_id, name, &mut self.send_pdu)?;
-        let recv_len = Self::send_and_recv(&self.socket, &self.send_pdu, &mut self.recv_buf[..])?;
+
+        let response = self.send_and_recv(&self.send_pdu)?;
+
+        handle_response(req_id, self.community.as_slice(), response.data.as_slice())
+    }
+
+    pub fn get_all_responses<T>(
+        &mut self,
+        names: &[T],
+        timeout: Duration,
+    ) -> SnmpResult<Vec<ResponseItem>>
+    where
+        T: AsRef<[u32]>,
+    {
+        let req_id = self.req_id.0;
         self.req_id += Wrapping(1);
-        let pdu_bytes = &self.recv_buf[..recv_len];
-        handle_response(req_id, self.community.as_slice(), pdu_bytes.into())
+
+        let socket = Self::create_socket(self.destination, Some(timeout), true)
+            .map_err(|_| SnmpError::SocketError)?;
+
+        // send
+        pdu::build_get(self.community.as_slice(), req_id, names, &mut self.send_pdu)?;
+        socket
+            .send_to(&self.send_pdu[..], self.destination)
+            .map_err(|_| SnmpError::SendError)?;
+
+        let ts1 = Instant::now();
+
+        // recv all responses
+        let mut vec1: Vec<ResponseItemInt> = Vec::new();
+        loop {
+            let response = Self::recv_one(&socket);
+            if response.is_err() || ts1.elapsed() >= timeout {
+                // skip any errors or delayed response
+                break;
+            }
+            vec1.push(response.unwrap());
+        }
+
+        // parsing to SnmpPdu
+        let mut vec2: Vec<ResponseItem> = Vec::new();
+        for item in vec1.iter() {
+            let r1 = handle_response(req_id, self.community.as_slice(), item.data.as_slice());
+            if r1.is_ok() {
+                vec2.push(ResponseItem {
+                    address: item.address.clone(),
+                    data: r1.unwrap(),
```

Comment:
Try to avoid unwraps. They will result in runtime errors. If a method returns `Result` use the ? operator instead.

Reply:
yes, in this case i checked the condition (line 198, r1.is_ok) before dereferencing it.

Reply:
In this case use `if let Ok(data) = r1 { ... }` or `match r1 { Ok(data) => ...`. please also use meaningful variable names without numbering.